torch/csrc/jit/codegen/cuda/scheduler/transpose.cpp

```diff
@@ -0,0 +1,616 @@
+  // [..., I1/tile1, .., I2/tile2, ..., tile1, tile2]
+
+  // Merge remaining dimensions
+  int lhs_i = -1;
+  for (int i = (int)reference1->nDims() - 2; i > 0; i--) {
+    auto axis_i = i - 1;
+    if (lhs_i == -1) {
+      lhs_i = axis_i;
+    } else {
+      reference1->merge(axis_i, lhs_i);
+      lhs_i = axis_i;
+    }
+  }
+  reference1->split(0, 1);
+  // [merged_dim, 1, tile1, tile2]
+
+  // parallelize non-tile dimensions
+  reference1->axis(1)->parallelize(ParallelType::Unswitch);
+  reference1->axis(0)->parallelize(ParallelType::BIDx);
+  // [BIDx, Unswitch, tile1, tile2]
+
+  // Propagate transformations so far to the entire DAG
+  TransformPropagator propagator(reference1);
+  MaxRootDomainInfoSpanningTree entire_dag(reference1);
+  entire_dag.traverse(&propagator);
+  scheduler_utils::parallelizeAllLike(reference1);
+
+  // For a transpose scheduling, all we need is to bind threadIdx.x differently
+  // for inputs and outputs. This swap of binding could happen at any tensor on
+  // the path from input to output, especially, it does not have to be in the
+  // transpose tensor. Here, we naively do the binding swap at cached
+  // input/output for simplicity. We might need to find a better set of swap
+  // tensors in the future to reduce shared memory usage.
+
+  // transform tile for vectorization/unroll
+  // See note [vectorization and unroll of input and output]
+
+  // schedule group 2
+  int pos = reference2->nDims() - 2;
+  // [..., tile1, tile2]
+  reference2->merge(pos);
+  reference2->split(pos, params.vectorize_factor2);
+  reference2->split(pos, kThreadsPerBlock);
+  // [..., Unroll, TIDx, Vectorize]
+
+  // Propagate transformations of reference2 to the entire DAG except
+  // group 1. We actually only want to propagate to the fusion outputs, but
+  // inputs and outputs themselves are disconnected, so we have to borrow the
+  // entire DAG and use its spanning tree.
+  {
+    auto all_tvs_except1 = ir_utils::allTvsExcept(
+        fusion,
+        {grouped_inputs_outputs[0].begin(), grouped_inputs_outputs[0].end()});
+    SetSelector selector({all_tvs_except1.begin(), all_tvs_except1.end()});
+    MaxRootDomainInfoSpanningTree entire_dag_except1(reference2, &selector);
+    TransformPropagator propagator(reference2);
+    entire_dag_except1.traverse(&propagator);
+  }
+
+  // parallelize group2 and its cached inputs
+  {
+    reference2->axis(-1)->parallelize(ParallelType::Vectorize);
+    reference2->axis(-2)->parallelize(ParallelType::TIDx);
+    reference2->axis(-3)->parallelize(ParallelType::Unroll);
+
+    ComputeAtMap ca_map(fusion);
+
+    scheduler_utils::parallelizeAllLike(
+        reference2,
+        {group2_and_cached_inputs.begin(), group2_and_cached_inputs.end()},
+        {ParallelType::Vectorize, ParallelType::TIDx});
+
+    // Only unrolled the axes that exactly maps to the unrolled axes
+    //  on reference as support for permissively mapped axes are not
+    //  yet clearly defined.
```

Comment:
I'm a little surprised, is it producing invalid results if you unroll non-exact mapped axes?

Reply:
Yes, exactly. This is the workaround @shmsong push.